peer/network.go

```diff
@@ -87,23 +89,25 @@ type network struct {
 	outstandingRequestHandlers map[uint32]message.ResponseHandler // maps avalanchego requestID => message.ResponseHandler
 	activeAppRequests          *semaphore.Weighted                // controls maximum number of active outbound requests
 	activeCrossChainRequests   *semaphore.Weighted                // controls maximum number of active outbound cross chain requests
-	appSender                  common.AppSender                   // avalanchego AppSender for sending messages
-	codec                      codec.Manager                      // Codec used for parsing messages
-	crossChainCodec            codec.Manager                      // Codec used for parsing cross chain messages
-	appRequestHandler          message.RequestHandler             // maps request type => handler
-	crossChainRequestHandler   message.CrossChainRequestHandler   // maps cross chain request type => handler
-	gossipHandler              message.GossipHandler              // maps gossip type => handler
-	peers                      *peerTracker                       // tracking of peers & bandwidth
-	appStats                   stats.RequestHandlerStats          // Provide request handler metrics
-	crossChainStats            stats.RequestHandlerStats          // Provide cross chain request handler metrics
+	router                     *p2p.Router
+	appSender                  common.AppSender                 // avalanchego AppSender for sending messages
+	codec                      codec.Manager                    // Codec used for parsing messages
+	crossChainCodec            codec.Manager                    // Codec used for parsing cross chain messages
+	appRequestHandler          message.RequestHandler           // maps request type => handler
+	crossChainRequestHandler   message.CrossChainRequestHandler // maps cross chain request type => handler
+	gossipHandler              message.GossipHandler            // maps gossip type => handler
+	peers                      *peerTracker                     // tracking of peers & bandwidth
+	appStats                   stats.RequestHandlerStats        // Provide request handler metrics
+	crossChainStats            stats.RequestHandlerStats        // Provide cross chain request handler metrics
 
 	// Set to true when Shutdown is called, after which all operations on this
 	// struct are no-ops.
 	closed utils.Atomic[bool]
 }
 
-func NewNetwork(appSender common.AppSender, codec codec.Manager, crossChainCodec codec.Manager, self ids.NodeID, maxActiveAppRequests int64, maxActiveCrossChainRequests int64) Network {
+func NewNetwork(router *p2p.Router, appSender common.AppSender, codec codec.Manager, crossChainCodec codec.Manager, self ids.NodeID, maxActiveAppRequests int64, maxActiveCrossChainRequests int64) Network {
```

Comment:
Following up on https://github.com/ava-labs/coreth/pull/266#discussion_r1313622140, how do we ensure that there are never overlaps between the requestID used by the SDK and the monotonically increasing requestID used by the current networking code?

I see you marked this comment as resolved on the old PR, how we avoid the following situation?

```
1. Issue a request via the SDK with requestID 1
2. Issue an unrelated request with the current networking code with requestID 1
3. Receive a response to the SDK request with requestID 1 and incorrectly clear the handler registered to the current networking code

Reply:
It seems like we treat an un-requested `AppResponse` as a fatal error in the p2p networking code assuming the engine ensures we don't receive unrequested responses, so I assume if we mess this up in the opposite order that it becomes an unexpected fatal error.

Reply:
I believe if we make two concurrent requests with the same request id, the last one to be sent out overwrites the previous one's request id ([ref](https://github.com/ava-labs/avalanchego/blob/14b8c98ce479d1c6f9c0fdb32f3fa60e6e137eb9/snow/networking/router/chain_router.go#L174)).

This means that the incoming `AppResponse` will only go to either the coreth handler, or the sdk handler.

In the case that the sdk request is taking over the codec request, this means that there will be a handler present from the original `AppRequest` flow, so we'll send the sdk response to the codec handler which will fail unmarshalling and we'll emit a debug log.

In the case that sdk goes first and codec goes second, then the codec handler will get its response as it expected and sdk will have its request time out correctly.

Reply:
> sdk will have its request time out correctly.

The SDK will never receive the timeout - as the request was cancelled.

Reply:
Ah yes. We might have to make an upstream change to fire AppRequestFailed on eviction then if there's a collision in the request id.

Reply:
See https://github.com/ava-labs/coreth/pull/316#discussion_r1317645330, we opted to break up the request id space into odd/even to avoid collisions. Turns out we do drop this and we don't fire the timeout.